### PR TITLE
tests/kernel/preempt: Add yield and sleep cases

### PR DESCRIPTION
**Pull request by andyross**
[Added test cases to try to catch the k_sleep() bug that Michael Scott and I were sure he had found in #8128.  Alas, this test PASSES, and I've now convinced myself that the bug wasn't real (though the symptoms are and there may be a similar issue elsewhere, just now with k_sleep()).  But they're useful regression cases and untested in current code, so worth commiting regardless]

Scheduler choice is subtle across yield and k_sleep(), add calls to
those to the state table and validate that we're making the right
decisions.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

### CODE REVIEW

File: tests/kernel/sched/preempt/src/main.c [hunk 1]
```diff
@@ -96,17 +106,21 @@ void wakeup_src_thread(int id)
 			     "worker thread %d not pending?", i);
 	}
 
+	/* Wake the src worker up */
 	last_thread = NULL;
 	k_sem_give(&worker_sems[id]);
 
+	while (do_sleep
+	       && !(worker_threads[id].base.thread_state & _THREAD_PENDING)) {
```
Comment (mike-scott) [inline tests/kernel/sched/preempt/src/main.c]: Won't this test immediately break when you submit the k_sleep re-write for v1.13?

**Review by mike-scott — CHANGES_REQUESTED** [review]

File: tests/kernel/sched/preempt/src/main.c (hunk 1 above)
Comment (andyross) [reply tests/kernel/sched/preempt/src/main.c]: Likely, yes.  :)  But I'm whiteboxing the thread states here anyway and this is the easiest way to make sure the thread is blocked on the semaphore again before going to the next test case.

Comment (mike-scott) [reply tests/kernel/sched/preempt/src/main.c]: Sounds good.  Removing my -1.

**Review by mike-scott — APPROVED** [review]
LGTM.  But we should track for updates in v1.13.

**Review by tyrantlizard77 — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Pull request merged**